cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -154,9 +180,26 @@ class CsvScanTask : public ScanTask {
         source_(fragment->source()) {}
 
   Result<RecordBatchIterator> Execute() override {
-    ARROW_ASSIGN_OR_RAISE(auto reader,
-                          OpenReader(source_, *format_, options(), options()->pool));
-    return IteratorFromReader(std::move(reader));
+    auto reader_fut = OpenReaderAsync(source_, *format_, options(),
+                                      internal::GetCpuThreadPool(), options()->pool);
+    auto reader_gen = GeneratorFromReader(std::move(reader_fut));
+    return MakeSimpleGeneratorIterator(std::move(reader_gen));
+  }
+
+  Future<RecordBatchVector> SafeExecute(internal::Executor* executor) override {
+    auto reader_fut =
+        OpenReaderAsync(source_, *format_, options(), executor, options()->pool);
```

Comment:
So `executor` here will be a SerialExecutor. Isn't that a behavior change? Previously, `StreamingReader::Make` would use the CPU thread pool. Now we're having it use the main thread.

Reply:
`SerialExecutor` assumes the current thread which may be from the CPU thread pool.  So, for example, `SyncScanner::ToTable` will continue to run `SafeExecute` on the CPU thread pool.

Reply:
Ahh, I see now - before, while a cpu_executor was passed around in StreamingReader, it wasn't actually used. (MakeStreamingReader accepts an executor, but never uses it.) Now it is used, but it'll be a SerialExecutor, so we remain single-threaded as desired.